File: scripts/matting.py
#!/usr/bin/env python3
"""
Triangulation Matting — Dual-pass alphamatte for 2D character sprites.

Background-removal algorithm that derives a true alpha channel from a
white-background + black-background image pair of the same subject.

Algorithm:
  White-bg pixel: C_white = α × C_fg + (1-α) × 1.0
  Black-bg pixel: C_black = α × C_fg + (1-α) × 0.0

  Subtract:  C_white - C_black = (1-α) × (1.0 - 0.0) = 1 - α
  Therefore: α = 1 - (C_white - C_black)

  Foreground: C_fg = C_black / α   (with safe division)

References:
  Smith & Blinn, "Blue Screen Matting" (SIGGRAPH 96)
  nano-banana-2-transparent (Replicate community model)

Usage:
  from scripts.matting import triangulation_matting
  triangulation_matting("parts/head_white.png", "parts/head_black.png", "parts/head.png")
"""

from pathlib import Path

try:
    from PIL import Image

    HAS_PIL = True
except ImportError:
    HAS_PIL = False

try:
    import numpy as np

    HAS_NUMPY = True
except ImportError:
    HAS_NUMPY = False


def triangulation_matting(white_path, black_path, output_path):
    """
    Produce an RGBA PNG with true alpha channel from a white-bg + black-bg pair.

    Args:
        white_path: Path to the white-background image (PNG).
        black_path: Path to the black-background image (PNG).
        output_path: Path to save the resulting RGBA PNG.

    Returns:
        True on success, False on failure.
    """
    if not HAS_PIL:
        print("❌ Pillow (PIL) is required. Install with: pip install Pillow")
        return False

    white_path = Path(white_path)
    black_path = Path(black_path)
    output_path = Path(output_path)

    if not white_path.exists():
        print(f"❌ White image not found: {white_path}")
        return False
    if not black_path.exists():
        print(f"❌ Black image not found: {black_path}")
        return False

    white = Image.open(white_path).convert("RGB")
    black = Image.open(black_path).convert("RGB")

    # Resize black to match white dimensions if needed
    if white.size != black.size:
        black = black.resize(white.size, Image.LANCZOS)

    result = _matting_numpy(white, black) if HAS_NUMPY else _matting_pure_pil(white, black)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    result.save(output_path, "PNG")
    return True


def _matting_numpy(white, black):
    """Fast path using numpy vectorized operations."""
    w_arr = np.array(white, dtype=np.float32) / 255.0
    b_arr = np.array(black, dtype=np.float32) / 255.0

    # α = 1 - (C_white - C_black) per channel, clamp to [0, 1]
    alpha_per_channel = 1.0 - (w_arr - b_arr)
    alpha_per_channel = np.clip(alpha_per_channel, 0.0, 1.0)

    # Use max alpha across RGB channels for robustness
    alpha = np.max(alpha_per_channel, axis=2)

    # C_fg = C_black / α
    safe_alpha = np.where(alpha > 0.005, alpha, 1.0)
    foreground = b_arr / safe_alpha[:, :, np.newaxis]
    foreground = np.clip(foreground, 0.0, 1.0)

    # Assemble RGBA
    h, w = alpha.shape
    result = np.zeros((h, w, 4), dtype=np.uint8)
    result[:, :, :3] = (foreground * 255).astype(np.uint8)
    result[:, :, 3] = (alpha * 255).astype(np.uint8)

    # Zero out fully-transparent pixels to avoid noise
    transparent_mask = alpha < 0.01
    result[transparent_mask, :] = 0

    return Image.fromarray(result, "RGBA")


def _matting_pure_pil(white, black):
    """Fallback path using pure Python + Pillow pixel iteration."""
    wpix = white.load()
    bpix = black.load()
    w, h = white.size

    result = Image.new("RGBA", (w, h), (0, 0, 0, 0))
    rpix = result.load()

    for y in range(h):
        for x in range(w):
            wr, wg, wb = wpix[x, y]
            br, bg, bb = bpix[x, y]

            # α per channel
            ar = max(0.0, min(1.0, 1.0 - (wr - br) / 255.0))
            ag = max(0.0, min(1.0, 1.0 - (wg - bg) / 255.0))
            ab = max(0.0, min(1.0, 1.0 - (wb - bb) / 255.0))

            # Use max alpha across channels
            alpha = max(ar, ag, ab)

            if alpha > 0.005:
                inv_a = 1.0 / alpha
                fr = max(0, min(255, int(br * inv_a)))
                fg = max(0, min(255, int(bg * inv_a)))
                fb = max(0, min(255, int(bb * inv_a)))
                fa = max(0, min(255, int(alpha * 255)))
                rpix[x, y] = (fr, fg, fb, fa)
            else:
                rpix[x, y] = (0, 0, 0, 0)

    return result


def has_matting_support():
    """Check if dependencies for matting are available."""
    return HAS_PIL

File: server/image/matting.ts
// Triangulation matting — port of matting.py (sharp).
//
// Recovers a straight-alpha transparent PNG from the same subject rendered on a
// white and a black background.
//
// Algorithm (per matting.py):
//   C_white = α·C_fg + (1-α)·1     C_black = α·C_fg + (1-α)·0
//   ⇒ C_white - C_black = 1 - α    ⇒ α = 1 - (C_white - C_black)
//   C_fg = C_black / α  (safe division)
// α is taken per-channel, clamped to [0,1], then max across RGB for robustness.
import sharp from "sharp";

/** Read an image as tightly-packed RGB (3 channels), returning pixels + size. */
async function toRgbRaw(png: Buffer): Promise<{ data: Buffer; width: number; height: number }> {
  const { data, info } = await sharp(png)
    .removeAlpha()
    .toColourspace("srgb")
    .raw()
    .toBuffer({ resolveWithObject: true });
  return { data, width: info.width, height: info.height };
}

/** Recover a transparent PNG from white-bg and black-bg renders of a subject. */
export async function triangulationMatting(whitePng: Buffer, blackPng: Buffer): Promise<Buffer> {
  const white = await toRgbRaw(whitePng);

  // Resize the black render to match the white render if dimensions differ.
  let blackBuf = blackPng;
  const blackMeta = await sharp(blackPng).metadata();
  if (blackMeta.width !== white.width || blackMeta.height !== white.height) {
    blackBuf = await sharp(blackPng)
      .resize(white.width, white.height, { fit: "fill", kernel: "lanczos3" })
      .png()
      .toBuffer();
  }
  const black = await toRgbRaw(blackBuf);

  const { width, height } = white;
  const w = white.data;
  const b = black.data;
  const out = Buffer.alloc(width * height * 4);

  const px = width * height;
  for (let i = 0; i < px; i++) {
    const s = i * 3;
    const wr = w[s]! / 255;
    const wg = w[s + 1]! / 255;
    const wb = w[s + 2]! / 255;
    const br = b[s]! / 255;
    const bg = b[s + 1]! / 255;
    const bb = b[s + 2]! / 255;

    // α = 1 - (C_white - C_black) per channel, clamped, then max across RGB.
    const ar = Math.min(1, Math.max(0, 1 - (wr - br)));
    const ag = Math.min(1, Math.max(0, 1 - (wg - bg)));
    const ab = Math.min(1, Math.max(0, 1 - (wb - bb)));
    const alpha = Math.max(ar, ag, ab);

    const o = i * 4;
    if (alpha < 0.01) {
      // Zero out fully-transparent pixels to avoid fringe noise.
      out[o] = 0;
      out[o + 1] = 0;
      out[o + 2] = 0;
      out[o + 3] = 0;
      continue;
    }

    const safe = alpha > 0.005 ? alpha : 1.0;
    const inv = 1 / safe;
    out[o] = Math.min(255, Math.max(0, Math.round(br * inv * 255)));
    out[o + 1] = Math.min(255, Math.max(0, Math.round(bg * inv * 255)));
    out[o + 2] = Math.min(255, Math.max(0, Math.round(bb * inv * 255)));
    out[o + 3] = Math.min(255, Math.max(0, Math.round(alpha * 255)));
  }

  return sharp(out, { raw: { width, height, channels: 4 } }).png().toBuffer();
}
